Feature/concurrent queue seqan

rrahn opened the pull request:
Blocked by ~#947~ and ~#948~

Supersedes #842 

Adds an initial version of the seqan2 concurrent queue. 
~This is still WIP and more tests will come.~
~Also there are some outstanding features that will be added in separate PRs after this is merged.~
~Note for me: make an issue with the remaining tasks.~

h-2 commented:
Without looking into this in detail: since you adapted SeqAn3 style guide and everything, why put this into contrib and not core/parallel or something like that?

rrahn commented:
Because there are more adaptions involved. I only adapted things that I had to touch anyway. Also the whole thing needs to be documented and everything, which is something we should have in the future but for now it will fit the purpose as is.

marehr commented on include/seqan3/contrib/parallel/concurrent_queue.hpp:
```diff
@@ -0,0 +1,478 @@
+        // increase capacity
+        data.resize(cap + 1);
+        size_type delta = data.size() - cap;
+        roundSize = static_cast<size_type>(1) << ((data.size() > 1) ? static_cast<size_type>(std::log2(data.size() - 1) + 1) : 1);
```
assert(delta == 1)?

marehr commented on test/unit/contrib/parallel/concurrent_queue_test.cpp:
```diff
@@ -0,0 +1,99 @@
+            t.join();
+    }
+
+    auto v = std::view::iota(0, size_v) | std::view::common;
```
small gauss

rrahn commented on include/seqan3/contrib/parallel/concurrent_queue.hpp:
```diff
@@ -0,0 +1,478 @@
+    size_type size_impl() const noexcept
+    {
+        size_type mask = roundSize - 1;
+        if ((headPos & mask) <= (tailPos & mask))
```
As discussed. Right now, we just moved it from SeqAn2 into contrib to get it working. Later we should put it into the library with all the documentation.

rrahn commented:
@marehr in the last commit I renamed the concurrent_queue to buffer_queue in order to be compliant with the standard proposal. I also added functionality for a fixed_size queue. So we have both a growing queue and a queue with fixed capacity.

rrahn commented:
jenkins was aborted.

marehr commented on test/unit/contrib/parallel/buffer_queue_base_test.cpp:
```diff
@@ -0,0 +1,138 @@
+
+using namespace seqan3;
+
+TEST(concurrent_queue, empty)
```
name `buffer_queue`

marehr commented on test/unit/contrib/parallel/buffer_queue_parallel_test.cpp:
```diff
@@ -0,0 +1,287 @@
+    EXPECT_TRUE(pop_status == queue_op_status::closed);
+}
+
+TEST(buffer_queue, spscx_dynamicsize)
```
xXx
